src/plugins/render/datatypes/texture.cpp

```diff
@@ -5,66 +5,135 @@
+			break;
+		case Texture::kGray:
+			TextureTraits<T_PTR>::init(GL_R8, width, height, data);
+			TextureTraits<T_PTR>::apply(GL_R, width, height, GL_RED, TextureTraits<T_PTR>::GLType(), data);
+			break;
+		default:
+			error = "Unsupported channel order for OpenGL texture storage.";
+			break;
 	}
 
-	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_BASE_LEVEL, 0);
-	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAX_LEVEL, 0);
-	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, format.interpolation);
-	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, format.interpolation);
+	glTexParameteri(TextureTraits<T_PTR>::BindType(), GL_TEXTURE_BASE_LEVEL, 0);
+	glTexParameteri(TextureTraits<T_PTR>::BindType(), GL_TEXTURE_MAX_LEVEL, 0);
+	glTexParameteri(TextureTraits<T_PTR>::BindType(), GL_TEXTURE_MIN_FILTER, format.interpolation);
+	glTexParameteri(TextureTraits<T_PTR>::BindType(), GL_TEXTURE_MAG_FILTER, format.interpolation);
 
-	glBindTexture(GL_TEXTURE_2D, 0);
+	glBindTexture(TextureTraits<T_PTR>::BindType(), 0);
 
 	glPixelStorei(GL_UNPACK_ALIGNMENT, original_alignment);
+
+	return error;
 }
 
-Texture::Texture(const float* data, std::size_t width, std::size_t height, const Format& format) : m_id(0) {
-	int original_alignment;
-	glGetIntegerv(GL_UNPACK_ALIGNMENT, &original_alignment);
-	glPixelStorei(GL_UNPACK_ALIGNMENT, format.row_byte_align);
+}
 
+Texture::Texture(const unsigned char* data, std::size_t width, std::size_t height, const Format& format) : m_id(0) {
 	glGenTextures(1, &m_id);
+	const std::string err = makeTexture(m_id, data, width, height, format);
 
-	glBindTexture(GL_TEXTURE_2D, m_id);
+	m_bindType = GL_TEXTURE_2D;
 
-	// this assumes all image data are stored as a flat float per channel array!
-	switch(format.channel_order) {
-		case kRGB: glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB32F, width, height, 0, GL_RGB, GL_FLOAT, data); break;
-		case kBGR: glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB32F, width, height, 0, GL_BGR, GL_FLOAT, data); break;
-		case kGray: throw std::runtime_error("Greyscale HDR images not supported."); break;
-		default: assert(false && "Invalid channel order");
+	if(!err.empty()) {
+		glDeleteTextures(1, &m_id);
+		throw std::runtime_error(err);
 	}
+}
 
+Texture::Texture(const float* data, std::size_t width, std::size_t height, const Format& format) : m_id(0) {
+	glGenTextures(1, &m_id);
+	const std::string err = makeTexture(m_id, data, width, height, format);
 
-	if(format.channel_order == kRGB)
-		glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB32F, width, height, 0, GL_RGB, GL_FLOAT, data);
-	else if(format.channel_order == kBGR)
-		glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB32F, width, height, 0, GL_BGR, GL_UNSIGNED_BYTE, data);
-	else
-		assert(false && "Invalid channel order");
+	m_bindType = GL_TEXTURE_2D;
 
+	if(!err.empty()) {
+		glDeleteTextures(1, &m_id);
+		throw std::runtime_error(err);
+	}
+}
+
+Texture::Texture(const std::vector<const unsigned char*> data, std::size_t width, std::size_t height, const Format& format) {
```

Comment:
Codacy found an issue: [Function parameter 'data' should be passed by const reference.](https://app.codacy.com/manual/martin.prazak/possumwood/pullRequest?prid=5900978)